File: frontend/src/api/execution.ts
import type { DataResponse, ListResponse } from "../types";
import { http } from "./client";

export interface Prompt {
  id: string;
  code: string;
  name: string;
  prompt_type: string;
  content: string;
  version: number;
  author: string | null;
}

export interface SOP {
  id: string;
  code: string;
  title: string;
  category: string;
  content: string;
  version: number;
}

export interface DecisionRule {
  id: string;
  ai_role_id: string;
  rule_type: string;
  name: string;
  description: string | null;
  authority_limit: string | null;
}

export interface QueueItem {
  id: string;
  ai_employee_id: string;
  ai_employee_name: string | null;
  work_item_id: string | null;
  work_item_code: string | null;
  title: string;
  description: string | null;
  priority: string;
  status: string;
  position: number;
  started_at: string | null;
  completed_at: string | null;
}

export interface HistoryEntry {
  id: string;
  ai_employee_id: string;
  ai_employee_name: string | null;
  work_item_id: string | null;
  action: string;
  output: string | null;
  status: string;
  duration_seconds: number | null;
  created_at: string;
}

export interface ReviewItem {
  id: string;
  work_item_id: string | null;
  work_item_code: string | null;
  reviewer_ai_employee_id: string;
  reviewer_name: string | null;
  submitter_ai_employee_id: string | null;
  submitter_name: string | null;
  status: string;
  notes: string | null;
  reviewed_at: string | null;
  created_at: string;
}

export interface Handoff {
  id: string;
  work_item_id: string | null;
  work_item_code: string | null;
  from_ai_employee_id: string | null;
  from_name: string | null;
  to_ai_employee_id: string;
  to_name: string | null;
  stage: string;
  status: string;
  notes: string | null;
  sequence: number;
}

export interface Workspace {
  employee: {
    id: string;
    employee_code: string;
    name: string;
    role: string | null;
    department: string | null;
  };
  context: string | null;
  inbox: Handoff[];
  assigned_tasks: {
    id: string;
    task_code: string;
    title: string;
    status: string;
    priority: string;
  }[];
  queue: QueueItem[];
  review_queue: ReviewItem[];
  history: HistoryEntry[];
  context_items: { key: string; value: string }[];
  kpis: { name: string; target: string | null; unit: string | null }[];
  performance: {
    queued: number;
    in_progress: number;
    completed: number;
    pending_reviews: number;
    avg_duration_seconds: number | null;
  };
}

export interface ExecFounderDash {
  ai_work_queue: number;
  queued: number;
  in_progress: number;
  pending_reviews: number;
  completed_work: number;
  avg_completion_seconds: number | null;
  organization_performance: { total_executions: number; handoffs: number };
}

export interface ExecAIDash {
  inbox: number;
  current_work: number;
  execution_queue: number;
  review_queue: number;
  history: number;
  work_by_employee: Record<string, number>;
}

export const executionApi = {
  prompts: () => http.get<ListResponse<Prompt>>("/prompts"),
  sops: () => http.get<ListResponse<SOP>>("/sops"),
  decisionRules: () => http.get<ListResponse<DecisionRule>>("/decision-rules"),
  workspace: (employeeId: string) => http.get<DataResponse<Workspace>>(`/workspaces/${employeeId}`),
  queue: (params: { aiEmployeeId?: string; status?: string } = {}) => {
    const q = new URLSearchParams();
    if (params.aiEmployeeId) q.set("ai_employee_id", params.aiEmployeeId);
    if (params.status) q.set("status", params.status);
    const qs = q.toString();
    return http.get<ListResponse<QueueItem>>(`/execution-queue${qs ? `?${qs}` : ""}`);
  },
  advanceQueue: (id: string, status: string, output?: string) =>
    http.post<DataResponse<QueueItem>>(`/execution-queue/${id}/advance`, { status, output }),
  history: () => http.get<ListResponse<HistoryEntry>>("/execution-history"),
  reviews: (status?: string) =>
    http.get<ListResponse<ReviewItem>>(`/reviews${status ? `?status=${status}` : ""}`),
  decideReview: (id: string, status: string, notes?: string) =>
    http.post<DataResponse<ReviewItem>>(`/reviews/${id}/decision`, { status, notes }),
  handoffs: (workItemId?: string) =>
    http.get<ListResponse<Handoff>>(`/handoffs${workItemId ? `?work_item_id=${workItemId}` : ""}`),
  founderDashboard: () => http.get<DataResponse<ExecFounderDash>>("/execution/founder-dashboard"),
  aiDashboard: () => http.get<DataResponse<ExecAIDash>>("/execution/ai-dashboard"),
};

File: backend/app/schemas/execution.py
"""Pydantic schemas for the AI Execution Engine."""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.domain.execution_enums import (
    DecisionRuleType,
    ExecutionStatus,
    HandoffStatus,
    PromptType,
    ReviewStatus,
    SOPCategory,
)
from app.domain.work_enums import Priority


# --- Prompt library -----------------------------------------------------
class PromptCreate(BaseModel):
    code: str = Field(min_length=1, max_length=60)
    name: str = Field(min_length=2, max_length=200)
    prompt_type: PromptType = PromptType.TASK
    content: str = Field(min_length=1)
    author: str | None = Field(default=None, max_length=160)


class PromptRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    code: str
    name: str
    prompt_type: PromptType
    content: str
    version: int
    author: str | None
    created_at: datetime
    updated_at: datetime


# --- SOP library --------------------------------------------------------
class SOPCreate(BaseModel):
    code: str = Field(min_length=1, max_length=60)
    title: str = Field(min_length=2, max_length=200)
    category: SOPCategory = SOPCategory.CODING
    content: str = Field(min_length=1)


class SOPRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    code: str
    title: str
    category: SOPCategory
    content: str
    version: int
    created_at: datetime
    updated_at: datetime


# --- Decision rules -----------------------------------------------------
class DecisionRuleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    ai_role_id: uuid.UUID
    rule_type: DecisionRuleType
    name: str
    description: str | None
    authority_limit: str | None


# --- Execution queue ----------------------------------------------------
class QueueItemCreate(BaseModel):
    ai_employee_id: uuid.UUID
    work_item_id: uuid.UUID | None = None
    title: str = Field(min_length=2, max_length=300)
    description: str | None = None
    priority: Priority = Priority.MEDIUM
    sop_id: uuid.UUID | None = None
    prompt_id: uuid.UUID | None = None


class QueueItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    ai_employee_id: uuid.UUID
    ai_employee_name: str | None = None
    work_item_id: uuid.UUID | None
    work_item_code: str | None = None
    title: str
    description: str | None
    priority: Priority
    status: ExecutionStatus
    position: int
    sop_id: uuid.UUID | None
    prompt_id: uuid.UUID | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class QueueAdvance(BaseModel):
    status: ExecutionStatus
    output: str | None = None


# --- Execution history --------------------------------------------------
class HistoryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    ai_employee_id: uuid.UUID
    ai_employee_name: str | None = None
    work_item_id: uuid.UUID | None
    action: str
    output: str | None
    status: ExecutionStatus
    duration_seconds: int | None
    created_at: datetime


# --- Review queue -------------------------------------------------------
class ReviewCreate(BaseModel):
    work_item_id: uuid.UUID | None = None
    reviewer_ai_employee_id: uuid.UUID
    submitter_ai_employee_id: uuid.UUID | None = None
    notes: str | None = None


class ReviewDecision(BaseModel):
    status: ReviewStatus
    notes: str | None = None


class ReviewRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    work_item_id: uuid.UUID | None
    work_item_code: str | None = None
    reviewer_ai_employee_id: uuid.UUID
    reviewer_name: str | None = None
    submitter_ai_employee_id: uuid.UUID | None
    submitter_name: str | None = None
    status: ReviewStatus
    notes: str | None
    reviewed_at: datetime | None
    created_at: datetime


# --- Handoffs -----------------------------------------------------------
class HandoffRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    work_item_id: uuid.UUID | None
    work_item_code: str | None = None
    from_ai_employee_id: uuid.UUID | None
    from_name: str | None = None
    to_ai_employee_id: uuid.UUID
    to_name: str | None = None
    stage: str
    status: HandoffStatus
    notes: str | None
    sequence: int
    created_at: datetime


class HandoffAdvance(BaseModel):
    status: HandoffStatus


# --- Execution context --------------------------------------------------
class ContextRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    ai_employee_id: uuid.UUID
    work_item_id: uuid.UUID | None
    key: str
    value: str
